### PR TITLE
feat(plugin): support return `sideEffect: 'no-treeshake'`

### PR DESCRIPTION
**Pull request by IWANABETHATGUY**
### Description
1. Part of ~https://github.com/rolldown/rolldown/issues/1266~

### CODE REVIEW

**Comment by hyf0** [comment]
#1266 is about https://rollupjs.org/configuration-options/#treeshake-modulesideeffects. I think you misunderstand it.

File: packages/rolldown/src/plugin/bindingify-build-hooks.ts [hunk 1]
```diff
@@ -149,28 +150,18 @@ export function bindingifyTransform(
...
-    }
-
-    if (ret.moduleSideEffects !== null) {
-      // @ts-ignore TODO The typing should import from binding
-      result.sideEffects = bindingifySideEffects(ret.moduleSideEffects)
-    }
-
-    return result
+      sideEffects: bindingifySideEffects(ret.moduleSideEffects),
+    } as unknown as BindingHookLoadOutput
```
Comment (hyf0) [inline packages/rolldown/src/plugin/bindingify-build-hooks.ts]: What's about this `as unknown`?  Code should be valid without it.

**Comment by IWANABETHATGUY** [comment]
> #1266 is about https://rollupjs.org/configuration-options/#treeshake-modulesideeffects. I think you misunderstand it.

I don't close the original issue, it is good to merge.

**Comment by hyf0** [comment]
> > #1266 is about https://rollupjs.org/configuration-options/#treeshake-modulesideeffects. I think you misunderstand it.
> 
> I don't close the original pull request, it is good to merge.

Wait a minue. I'm fixing the `as unknown`.

File: crates/rolldown/src/stages/link_stage/tree_shaking.rs [hunk 2]
```diff
@@ -23,16 +23,23 @@ fn include_module(ctx: &mut Context, module: &NormalModule) {
...
+    // runtime module has no side effects and it's statements should be included
+    // by other modules's references.
+    return;
+  }
+
+  if ctx.tree_shaking {
+    let forced_no_treeshake = matches!(module.side_effects, DeterminedSideEffects::NoTreeshake);
     module.stmt_infos.iter_enumerated().for_each(|(stmt_info_id, stmt_info)| {
-      if stmt_info.side_effect {
+      if stmt_info.side_effect && !forced_no_treeshake {
```
Comment (IWANABETHATGUY) [inline crates/rolldown/src/stages/link_stage/tree_shaking.rs]: Any good of this? you need to calculate every time, we only calculate once previously.

Comment (hyf0) [reply crates/rolldown/src/stages/link_stage/tree_shaking.rs]: I want make it more clear to see the different between enabling treeshaking or not. But yes bad for performance. Let's do a better way.

**Review by hyf0 — APPROVED** [review]
@IWANABETHATGUY Done. Leave to you if there are still things need to be changed.

**Comment by hyf0** [comment]
@IWANABETHATGUY Sorry to overwrite your commit. I will do MR next time.

**Comment by hyf0** [comment]
I need to change some code in `normal_module_task.rs`. To prevent conflicts, let me merge the PR first. If there are still things need to be changed, let's do them in another PR.

**Comment by IWANABETHATGUY** [comment]
> @IWANABETHATGUY Sorry to overwrite your commit. I will do MR next time.

It is alright,  I pushed the patch before I saw your comments, :rofl:

**Pull request merged**